src/core/vectortile/qgsvectortilewriter.cpp

```diff
@@ -115,32 +149,85 @@ bool QgsVectorTileWriter::writeTiles( QgsFeedback *feedback )
           continue;
         }
 
-        QString filePath = QgsVectorTileUtils::formatXYZUrlTemplate( sourcePath, tileID );
-
-        // make dirs if needed
-        QFileInfo fi( filePath );
-        QDir fileDir = fi.dir();
-        if ( !fileDir.exists() )
+        if ( sourceType == QStringLiteral( "xyz" ) )
         {
-          if ( !fileDir.mkpath( "." ) )
-          {
-            mErrorMessage = tr( "Cannot create directory " ) + fileDir.path();
-            return false;
-          }
+          if ( !writeTileFileXYZ( sourcePath, tileID, tileData ) )
+            return false;  // error message already set
         }
-
-        QFile f( filePath );
-        if ( !f.open( QIODevice::WriteOnly ) )
+        else  // mbtiles
         {
-          mErrorMessage = tr( "Cannot open file for writing " ) + filePath;
-          return false;
+          QByteArray gzipTileData;
+          QgsMBTilesReader::encodeGzip( tileData, gzipTileData );
+          int rowTMS = pow( 2, tileID.zoomLevel() ) - tileID.row() - 1;
+          mbtiles->setTileData( tileID.zoomLevel(), tileID.column(), rowTMS, gzipTileData );
         }
-
-        f.write( tileData );
-        f.close();
       }
     }
   }
 
   return true;
 }
+
+bool QgsVectorTileWriter::writeTileFileXYZ( const QString &sourcePath, QgsTileXYZ tileID, const QByteArray &tileData )
+{
+  QString filePath = QgsVectorTileUtils::formatXYZUrlTemplate( sourcePath, tileID );
+
+  // make dirs if needed
+  QFileInfo fi( filePath );
+  QDir fileDir = fi.dir();
+  if ( !fileDir.exists() )
+  {
+    if ( !fileDir.mkpath( "." ) )
+    {
+      mErrorMessage = tr( "Cannot create directory " ) + fileDir.path();
+      return false;
+    }
+  }
+
+  QFile f( filePath );
+  if ( !f.open( QIODevice::WriteOnly ) )
+  {
+    mErrorMessage = tr( "Cannot open file for writing " ) + filePath;
+    return false;
+  }
+
+  f.write( tileData );
+  f.close();
+  return true;
+}
+
+QString QgsVectorTileWriter::mbtilesJsonSchema()
+{
+  QJsonArray arrayLayers;
```

Comment:
I think the general preference in the code seems to be moving away from QJson -> nlohmann. You could replace this with QgsJsonUtils::jsonFromVariant , e.g.

    return QString::fromStdString( QgsJsonUtils::jsonFromVariant( value ).dump( 2 ) );